### PR TITLE
Add duality-gap thresholds

### PR DESCRIPTION
**Pull request by stephane-caron**
Working draft accompanying the discussion in https://github.com/Simple-Robotics/proxsuite/issues/168

Closes https://github.com/Simple-Robotics/proxsuite/issues/168

### CODE REVIEW

**Comment by hrp2-14** [comment]
Hi ! This project doesn't usually accept pull requests on the main branch.
If this wasn't intentionnal, you can change the base branch of this PR to devel
(No need to close it for that). Best, a bot.

File: include/proxsuite/proxqp/sparse/solver.hpp [hunk 1]
```diff
@@ -844,7 +844,7 @@ qp_solve(Results<T>& results,
           is_dual_feasible(dual_feasibility_lhs)) {
         if (settings.check_duality_gap) {
           if (std::abs(results.info.duality_gap) <=
```
Comment (stephane-caron) [inline include/proxsuite/proxqp/sparse/solver.hpp:846]: We use both ``std::abs`` and ``std::fabs`` in the four instances of this check.

AFAIU the latter is float-specific and the former more general, so we should consistently use ``std::fabs``, right?

File: bindings/python/src/expose-settings.hpp [hunk 2]
```diff
@@ -71,6 +71,8 @@ exposeSettings(pybind11::module_ m)
                    &Settings<T>::compute_preconditioner)
     .def_readwrite("update_preconditioner", &Settings<T>::update_preconditioner)
     .def_readwrite("check_duality_gap", &Settings<T>::check_duality_gap)
+    .def_readwrite("eps_gap_abs", &Settings<T>::eps_gap_abs)
+    .def_readwrite("eps_gap_rel", &Settings<T>::eps_gap_rel)
```
Comment (jcarpent) [inline bindings/python/src/expose-settings.hpp:75]: Would be better to rename it fully as follows:
```suggestion
    .def_readwrite("eps_duality_gap_abs", &Settings<T>::eps_gap_abs)
    .def_readwrite("eps_duality_gap_rel", &Settings<T>::eps_gap_rel)
```

**Review by jcarpent — CHANGES_REQUESTED** [review]
Thanks @stephane-caron for the PR.
I would just recommend changing the choice of variable naming to use the full name `duality_gap` for clarity and transparency.

File: include/proxsuite/proxqp/sparse/solver.hpp (hunk 1 above)
Comment (jcarpent) [reply include/proxsuite/proxqp/sparse/solver.hpp:846]: Yes, this is a mistake ...

Comment (stephane-caron) [reply include/proxsuite/proxqp/sparse/solver.hpp]: OK, done in 2d7a82bae402a2fda422fe2f671c56c1a140c89e.

File: bindings/python/src/expose-settings.hpp (hunk 2 above)
Comment (stephane-caron) [reply bindings/python/src/expose-settings.hpp]: Done in b480bc26c1d14b0d4c4a226968327f2e2c9a95d2

**Comment by stephane-caron** [comment]
Ready for review! The main difference since last review is the proposal to replace the instructions:

```
-  duality_gap /= sqrt_max_dim; // in order to get an a-dimensional duality gap
-  rhs_duality_gap /= sqrt_max_dim;
```

By a recommendation in the docs:

> ... contrary to residuals which result from an infinite norm, the duality gap scales with the square root of the problem dimension (thus it is numerically harder to achieve a given duality gap for larger problems). A recommended choice is $\epsilon^{\text{gap}}\_{\text{abs}} = \epsilon_{\text{abs}} \sqrt{\max(n, n_{\text{eq}}, n_{\text{ineq}})}$.

**Pull request merged**